src/bootstrap/cc_detect.rs

```diff
@@ -120,12 +121,12 @@ pub fn find(build: &mut Build) {
         // We'll need one anyways if the target triple is also a host triple
         let mut cfg = new_cc_build(build, target);
         cfg.cpp(true);
-        let cxx_configured = if let Some(cxx) = config.and_then(|c| c.cxx.as_ref()) {
+        let cxx_configured = if let Some(cxx) = config
+            .and_then(|c| c.cxx.clone())
+            .or_else(|| default_compiler(&mut cfg, Language::CPlusPlus, target, build))
+        {
             cfg.compiler(cxx);
             true
-        } else if build.hosts.contains(&target) || build.build == target {
```

Comment:
I'm trying to confirm whether the removal of this if condition (to default_compiler above) is intentional - it seems like it would change behavior for at least some targets?

Reply:
Yes, it's needed to keep the tests passing on Android, because it turns out that `$CXX` needs to be set correctly for the `issue-36710` test to pass. If I bring back something like the original logic like so:
```
diff --git a/src/bootstrap/cc_detect.rs b/src/bootstrap/cc_detect.rs
index 7b84e990ae5..714ec6284c1 100644
--- a/src/bootstrap/cc_detect.rs
+++ b/src/bootstrap/cc_detect.rs
@@ -125,8 +125,12 @@ pub fn find(build: &mut Build) {
             .and_then(|c| c.cxx.clone())
             .or_else(|| default_compiler(&mut cfg, Language::CPlusPlus, target, build))
         {
-            cfg.compiler(cxx);
-            true
+            if build.hosts.contains(&target) || build.build == target {
+                cfg.compiler(cxx);
+                true
+            } else {
+                cfg.try_get_compiler().is_ok()
+            }
         } else {
             // Use an auto-detected compiler (or one configured via `CXX_target_triple` env vars).
             cfg.try_get_compiler().is_ok()

```
And run the tests like so (with #102757 patched in to make the `std` tests pass):
```
python3 x.py test --target aarch64-linux-android --exclude src/tools/linkchecker --exclude tests/debuginfo --force-rerun
```
I see a failure later:
```
--- stdout -------------------------------
aarch64-linux-android-clang++ -DANDROID -ffunction-sections -fdata-sections -fPIC -c -o /mnt/disk2/pcc/rust3/build/x86_64-unknown-linux-gnu/test/run-make/issue-36710/issue-36710/libfoo.o foo.cpp
------------------------------------------
--- stderr -------------------------------
/bin/dash: 1: aarch64-linux-android-clang++: not found
make: *** [Makefile:17: /mnt/disk2/pcc/rust3/build/x86_64-unknown-linux-gnu/test/run-make/issue-36710/issue-36710/libfoo.o] Error 127
------------------------------------------
```
This change makes the compiler detection for C++ consistent with C, which also first tries the logic in `default_compiler()` to find the compiler in all cases and then tries `cc::Build::get_base_compiler()` via `try_get_compiler()` if that fails. Previously for C++ the `default_compiler()` (previously `set_compiler()`) call would be omitted for targets that are cross-compiling the standard library, leading to the test failure on Android.

I think the circumstances where this change would break a build are:
1. When cross-compiling the standard library, `get_base_compiler()` returns a working C++ compiler but `default_compiler()` returns a broken one.
2. When cross-compiling the standard library and `get_base_compiler()` returns `None` for the C++ compiler for the target, causing the `get_compiler()` call in `default_compiler()` to fail, aborting the `bootstrap` program.

Looking at the individual non-Android cases:
- `openbsd`: in some cases replaces `gcc` with `egcc`, as well as replacing `g++` with `eg++`. This seems right, there [appears to be](https://google.com/search?q=openbsd+"eg%2B%2B") a compiler named `eg++` on some versions of OpenBSD. So I suspect this could actually fix (to some degree) cross-compiling to OpenBSD if `eg++` is available as a cross compiler. OpenBSD has a C++ compiler in the base system, so the call to `get_compiler()` seems okay.
- `mips*-unknown-linux-musl` replaces `gcc` with `mips*-linux-musl-gcc`, which will have no effect in the C++ case. Other musl targets replace the compiler with `musl-gcc` regardless of language. This isn't right for C++ and there doesn't seem to be such a thing as `musl-g++` in a normal musl installation (at least the musl sources don't mention it) and it seems plausible that a musl installation would support C but not C++, so `get_base_compiler()` could fail. The right fix seems to be to guard all of the `musl` clauses with a check that `compiler == Language::C`. That means we will fall back to `try_get_compiler()` for C++ musl targets, as we were doing before.

I've made the proposed fix for `musl` targets in the latest update of the PR.